Source/Core/Core/PowerPC/Jit64/Jit64_Util.cpp

```diff
@@ -0,0 +1,420 @@
+  else
+  {
+    if (swap)
+      SwapAndStore(accessSize, target, arg.GetSimpleReg());
+    else
+      MOV(accessSize, target, arg);
+  }
+}
+
+void Jit64::UnsafeWriteGatherPipe(int accessSize, GPRRegister& value, bool swap)
+{
+  u32 gather_pipe = (u32)(u64)GPFifo::m_gatherPipe;
+  _assert_msg_(DYNA_REC, gather_pipe <= 0x7FFFFFFF, "Gather pipe not in low 2GB of memory!");
+
+  // Append this write to the pipe
+  auto scratch = regs.gpr.Borrow();
+  MOV(32, R(scratch), M(&GPFifo::m_gatherPipeCount));
+  UnsafeWriteToMemory(accessSize, value, MDisp(scratch, gather_pipe), swap);
+  ADD(32, R(scratch), Imm8(accessSize >> 3));
+  MOV(32, M(&GPFifo::m_gatherPipeCount), R(scratch));
+  jit->js.fifoBytesThisBlock += accessSize >> 3;
+}
+
+void Jit64::UnsafeWriteToConstRamAddress(int accessSize, GPRRegister& value, u32 address, bool swap)
+{
+  // We can avoid a scratch register if the address is in the lower half
+  if (address > 0x80000000)
+  {
+    // Technically the borrow doesn't last long enough, but nothing will
+    // re-allocate it before we're done
+    auto addr = regs.gpr.Borrow();
+    MOV(32, addr, Imm32(address));
+    UnsafeWriteToMemory(accessSize, value, MRegSum(RMEM, addr), swap);
+  }
+  else
+  {
+    UnsafeWriteToMemory(accessSize, value, MDisp(RMEM, address), swap);
+  }
+}
+
+void Jit64::SafeLoad(GPRNative& reg_value, GPRRegister& reg_addr, GPRRegister& offset,
+                     int accessSize, bool signExtend, bool swap, bool update)
+{
+  bool valueAddressShareRegister = reg_value.IsAliasOf(reg_addr);
+  bool valueOffsetShareRegister = reg_value.IsAliasOf(offset);
+
+  if (update && !reg_addr.HasPPCRegister())
+  {
+    // This form of instruction is illegal
+    WriteInvalidInstruction();
+    return;
+  }
+
+  if (!swap && accessSize == 8)
+  {
+    // Swapping automatically is currently cheaper
+    swap = true;
+  }
+
+  // Constant writes have a much faster path
+  if (reg_addr.IsImm() && offset.IsImm())
+  {
+    u32 const_addr = reg_addr.Imm32() + offset.Imm32();
+    if (UnsafeReadFromConstAddress(accessSize, reg_value, const_addr, signExtend, swap))
+    {
+      if (update)
+        reg_addr.SetImm32(const_addr);
+      return;
+    }
+
+    // Nope, we'll fall back to the standard read...
+  }
+
+  if (!swap)
+  {
+    // TODO: add a flag to disable swap in SafeLoadToReg
+    SafeLoad(reg_value, reg_addr, offset, accessSize, signExtend, true, update);
+    BSWAP(accessSize, reg_value);
+    return;
+  }
+
+  // Easy: zero offset, so updates don't matter
+  if (offset.IsZero())
+  {
+    // Neither of these is transactional, since we won't update address
+    // unless the load succeeds
+    if (valueAddressShareRegister)
+    {
+      auto xv = reg_value.Bind(BindMode::ReadWrite);
+      SafeLoadToReg(xv, xv, accessSize, 0, CallerSavedRegistersInUse(), signExtend, 0);
+    }
+    else
+    {
+      auto addr =
+          (reg_addr.IsImm() || reg_addr.IsRegBound()) ? reg_addr : reg_addr.Bind(BindMode::Read);
```

Comment:
It seems like you use this pattern all over the place; maybe you want a BindMode::ReadOrImm?